### PR TITLE
[4.0] [plg_system_httpheader] Implement security http headers per default and create the posibility to add custom http headers

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

Implement security http headers per default and create the possibility to add custom http headers.

This plugin bases on https://github.com/zero-24/plg_system_httpheader which does the same for 3.x

### Testing Instructions

- test your webseite using: https://securityheaders.io/ => `The most if not all headers are missing`
- Install this PR.
- discover the new plugin
- check that the plugin is enabled if not enable it.
- check the report here: https://securityheaders.io/ again.

You can also check the http header using the `Network` Tab of your Browser:
image

Per default this plugin set the following headers + values:
```
X-Frame-Options: SAMEORIGIN
```
More Infos: https://scotthelme.co.uk/hardening-your-http-response-headers/#x-frame-options
```
X-XSS-Protection: 1; mode=block
```
More Infos: https://scotthelme.co.uk/hardening-your-http-response-headers/#x-xss-protection
```
X-Content-Type-Options: nosniff
```
More Infos: https://scotthelme.co.uk/hardening-your-http-response-headers/#x-content-type-options
```
Referrer-Policy: no-referrer-when-downgrade
```
More Infos: https://scotthelme.co.uk/a-new-security-header-referrer-policy/

### Expected result

After enabeling the default options the above mention headers are set.

### Actual result

Joomla does not set any security headers.

### Additional Infos

Give of the nature of the other headers we can not give a save default as they are very site specific. But atleast this plugin offers a easy to implement way to add the other headers using subforms.

Thanks for @SniperSister @wilsonge and @yvesh for the feedback on the first implementation ;)

### Next todo's

- [ ] Please test the plugin
- [x] We should review the language strings
- [x] We need to decide if the descriptions can stay or should be dropped. 
- [ ] Give your feedback.

Thanks.

### CODE REVIEW

File: administrator/language/en-GB/en-GB.plg_system_httpheader.sys.ini [hunk 1]
```diff
@@ -0,0 +1,32 @@
+; Joomla! Project
+; Copyright (C) 2005 - 2017 Open Source Matters. All rights reserved.
+; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
+; Note : All ini files need to be saved as UTF-8
+
+PLG_SYSTEM_HTTPHEADER="System - HttpHeader"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheader.sys.ini]: HTTP

**Comment by brianteeman** [comment]
If we are going to do this then we should be complete and also offer CSP and STS options.

File: plugins/system/httpheader/httpheader.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	/**
+	 * Application object.
+	 *
+	 * @var    JApplicationCms
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected $app;
+
+	/**
+	 * The list of the suported HTTP headers
```
Comment (Quy) [inline plugins/system/httpheader/httpheader.php]: change to `supported `

File: plugins/system/httpheader/httpheader.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	 */
+	public static function getSubscribedEvents()
+	{
+		return [
+			'onAfterInitialise' => 'setHttpHeaders',
+		];
+	}
+
+	/**
+	 * The `setHttpHeaders` methode handle the setting of the configured HTTP Headers
```
Comment (Quy) [inline plugins/system/httpheader/httpheader.php]: Fix `methode`

**Comment by zero-24** [comment]
> If we are going to do this then we should be complete and also offer CSP and STS options.

hmm the problem with that both things is that you can not set just one default option. But anleast for the csp there are planings beyond this PR / Plugin by the Joomla 4 team. So the system can generate a suggestion. But this should not stop this plugin as this prepare anything we can do from now.

Thanks for your other feedback. Commit is incoming 😄

**Comment by brianteeman** [comment]
Agree about the issue of the default. Although it is possible to have a default value that will pass the test (I've done it) it is a crazy value

**Comment by wilsonge** [comment]
@zero-24 do you want to add STS header into this PR - i think that one is pretty easy

@brianteeman i'm going to leave CSP for a later addition to this plugin - as you know from my experience at the sprint with https://github.com/wilsonge/joomla-cms/tree/csp it was lots of pain to have a half decent CSP - i'm going to keep plugging away at it. But in the mean time I think this gives us some quick wins whilst we think about things

**Comment by brianteeman** [comment]
Re the descriptions. All the descriptions that are just long winded ways of saying the same thing as the label can be removed.

**Comment by zero-24** [comment]
> @zero-24 do you want to add STS header into this PR - i think that one is pretty easy

Sure but what default value do you suggest? And as we don't force https on new sites it need to be per default disabled anyway.

**Comment by zero-24** [comment]
>  All the descriptions that are just long winded ways of saying the same thing as the label can be removed.

Sure. The question is what are our target users. Does a "normal" enduser understand what the setting means or not?

**Comment by wilsonge** [comment]
> Sure but what default value do you suggest? And as we don't force https on new sites it need to be per default disabled anyway.

I think we don't need on/off - we can use the main application values for force https? All we need to do is have the time till expires in the plugin which is configurable (by default probably a month - 2592000 seconds)

**Comment by zero-24** [comment]
> I think we don't need on/off - we can use the main application values for force https? All we need to do is have the time till expires in the plugin which is configurable (by default probably a month - 2592000 seconds)

Please check https://github.com/joomla/joomla-cms/pull/18301/files#diff-660b387da9d45a7ecca4cd27e25b3280R126 is it that what you had in mind?

Should we include the `includeSubDomains` setting too?

I have skipped the option part and allow now to overwrite the values using the `additional headers` i think this keeps the options low.

**Comment by wilsonge** [comment]
> Please check https://github.com/joomla/joomla-cms/pull/18301/files#diff-660b387da9d45a7ecca4cd27e25b3280R126 is it that what you had in mind?

Seems good to me!

> Should we include the `includeSubDomains` setting too?

Umm I wouldn't do `includeSubDomains` - we're only concerned about the current Joomla site - no more

**Comment by SniperSister** [comment]
If i read this correctly, the default headers will be enabled even for updated 3.x installs, correct? If so, we should all have in mind that this might be experienced as a backwards breaking change for users that are running their site in an iframe (I know a number of hosts who implement "redirects where the original domain stays the same" using an iframe) and make sure that this is documented properly.

**Comment by brianteeman** [comment]
PLG_SYSTEM_HTTPHEADER_XFRAMEOPTIONS_DESC and the following descriptions are great examples of !eaningless strings. They don't say anything. And if your argument is that they are useful for less technical users then they are worse than useless as they are without any useful information (which should be in docs not tooltips anyway)

**Comment by brianteeman** [comment]
Default on will be a problem. No question about that.

**Comment by brianteeman** [comment]
https://github.com/joomla/joomla-cms/pull/18301#issuecomment-335630499 no we can't use the force HTTPS value in global config. On a good server you do not (and should not) use that

**Comment by wilsonge** [comment]
@SniperSister you didn't because the PR is against 4.0 branch :)

**Comment by SniperSister** [comment]
@wilsonge sure, but 3.x sites will be updated to 4.x in the future too? ;)

**Comment by brianteeman** [comment]
Security headers are great and its great that we will have a way to implement them in joomla.

However they really are not something that should be enabled by default - even with what you think are sensible settings. Especially as because when security headers block something from loading you do not get an error message indicating where the issue is. 

If you think of the content security we have right now with creating xframes in the weysiwyg editor and how many issues this creates on the forum - and now massively multiply it you should get the idea.

I've been fortunate to meet and hear Scott speak a few times and I am confident he would agree that security headers should be something that you implement on your web site consciously from a position of knowledge and understanding.

For example one of my sites has security headers and an A rating yet it also has a policy to allow in-line styles which is kind of crazy - but that was my informed choice - otherwise my site would have broken.

**Comment by zero-24** [comment]
For the default enabled issue we can go the postinstall message for upgrades and default enabled for new installs.
Can we agree on that? I can build it than like this.

For the descriptions i can drop them.

Sure the csp's need to be set on a informed choice. This is the reason i have not implemented a default value and told above that this (suggesting a value) is out of scope for now ;-)

As for the HSTS imo the same applys but i have implemented the poc based on georges comment. For me we can and should drop that default too as this should be a informed choice too. As this also comes with consequences.

**Comment by brianteeman** [comment]
We could have the plugin default enabled with all the default values being empty and a post installation message. This will people the time to understand and make informed choices without breaking their sites. Otherwise we are making everyone do that first before testing their shiny new J4 site which we know they will want to do BEFORE looking at messages - especially hard ones they won't understand and where the best resource for information, Scott's site, is in English only.

File: administrator/language/en-GB/en-GB.plg_system_httpheader.sys.ini [hunk 4]
```diff
@@ -0,0 +1,32 @@
+; Joomla! Project
+; Copyright (C) 2005 - 2017 Open Source Matters. All rights reserved.
+; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
+; Note : All ini files need to be saved as UTF-8
+
+PLG_SYSTEM_HTTPHEADER="System - HTTPHeader"
+PLG_SYSTEM_HTTPHEADER_XML_DESCRIPTION="This Plugin can set some Security HTTP Header"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheader.sys.ini]: Headers

File: plugins/system/httpheader/httpheader.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Plugin\CMSPlugin;
+use Joomla\Event\SubscriberInterface;
+
+/**
+ * Plugin class for Http Header
```
Comment (brianteeman) [inline plugins/system/httpheader/httpheader.php]: HTTP

File: plugins/system/httpheader/httpheader.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+ *
+ * @since  __DEPLOY_VERSION__
+ */
+class PlgSystemHttpHeader extends CMSPlugin implements SubscriberInterface
+{
+	/**
+	 * Affects constructor behavior. If true, language files will be loaded automatically.
```
Comment (brianteeman) [inline plugins/system/httpheader/httpheader.php]: as this is documentations it should be in en-GB so "behaviour"

File: plugins/system/httpheader/httpheader.php [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	public function setHttpHeaders()
+	{
+		$this->setDefaultHeader();
+
+		// Handle the additional httpheader
+		$httpHeaders = $this->params->get('additional_httpheader', array());
+
+		foreach ($httpHeaders as $httpHeader)
+		{
+			// Handle the client settings foreach header
```
Comment (brianteeman) [inline plugins/system/httpheader/httpheader.php]: for each is two words

File: plugins/system/httpheader/httpheader.php (hunk 5 above)
Comment (brianteeman) [reply plugins/system/httpheader/httpheader.php]: being really pedantic but when I see something like this `PlgSystemHttpHeader ` I want to change it to `PlgSystemHTTPHeader `

File: plugins/system/httpheader/postinstall/introduction.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+ * @since   4.0
+ */
+function httpheader_postinstall_action()
+{
+	// Enable the plugin
+	$db = JFactory::getDbo();
+
+	$query = $db->getQuery(true)
+		->update($db->qn('#__extensions'))
+		->set($db->qn('enabled') . ' = ' . $db->q(1))
```
Comment (alikon) [inline plugins/system/httpheader/postinstall/introduction.php]: from `#__extensions` table  the `enabled` field is not declared as a `string`
`enabled tinyint(3) NOT NULL DEFAULT 0,`

File: administrator/language/en-GB/en-GB.plg_system_httpheader.sys.ini [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+PLG_SYSTEM_HTTPHEADER_XML_DESCRIPTION="This Plugin can set some Security HTTP Headers"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER="Define additional HTTP Header"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_DESC="Using this you can set also different values from the default ones and also additional headers when it is supported. The supported headers are: <br><ul><li>Strict-Transport-Security</li><li>Content-Security-Policy</li><li>Content-Security-Policy-Report-Only</li><li>X-Frame-Options</li><li>X-XSS-Protection</li><li>X-Content-Type-Options</li><li>Referrer-Policy</li><li>Expect-CT</li></ul>"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_CLIENT="Site selection"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_CLIENT_SITE="Site"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_CLIENT_ADMINISTRATOR="Administrator"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_CLIENT_BOTH="Both"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_KEY="HTTP Header"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_VALUE="HTTP Header Value"
+PLG_SYSTEM_HTTPHEADER_POSTINSTALL_INTRODUCTION_TITLE="HTTP Security Headers are available"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheader.sys.ini]: This doesnt seem correct I would delete the "are available"

File: administrator/language/en-GB/en-GB.plg_system_httpheader.sys.ini [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
...><a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-frame-options'>'X-Frame-Options: SAMEORIGIN'</a></li><li><a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-xss-protection'>'X-XSS-Protection: 1; mode=block'</a></li><li><a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-content-type-options'>'X-Content-Type-Options: nosniff'</a></li><li><a href='https://scotthelme.co.uk/a-new-security-header-referrer-policy/'>'Referrer-Policy: no-referrer-when-downgrade'</a></li></ul><br><br>The full list of supported headers are: <br><ul><li><a href='https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security'>Strict-Transport-Security</a></li><li><a href='https://en.wikipedia.org/wiki/Content_Security_Policy'>Content-Security-Policy</a></li><li>Content-Security-Policy-Report-Only</li><li>X-Frame-Options</li><li>X-XSS-Protection</li><li>X-Content-Type-Options</li><li>Referrer-Policy</li><li>Expect-CT</li></ul><br><br>These headers help your browser to protect your website from <a href='https://en.wikipedia.org/wiki/Cross-site_scripting'>XSS</a> and <a href='https://en.wikipedia.org/wiki/Clickjacking'>Clickjacking</a> attacks."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheader.sys.ini]: handles 
It helps to secure your site

File: plugins/system/httpheader/httpheader.php (hunk 3 above)
Comment (brianteeman) [reply plugins/system/httpheader/httpheader.php]: To be clear this should be method without an e at the end

File: administrator/language/en-GB/en-GB.plg_system_httpheader.sys.ini [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+PLG_SYSTEM_HTTPHEADER_POSTINSTALL_INTRODUCTION_ACTION="Enable default security headers"
+PLG_SYSTEM_HTTPHEADER_REFERRERPOLICY="Referrer-Policy"
+PLG_SYSTEM_HTTPHEADER_REFERRERPOLICY_EMPTY_STRING="Empty String"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheader.sys.ini]: > Empty String

> An empty string value in the Referrer Policy header indicates that the site doesn't want to set a Referrer Policy here and the browser should fallback to a Referrer Policy defined via other mechanisms elsewhere

So in this case I think it is better to have the string be "disabled" or "not set"

File: plugins/system/httpheader/httpheader.php [hunk 12]
```diff
@@ -123,12 +123,6 @@ private function setDefaultHeader()
 			$this->app->setHeader('X-Content-Type-Options', 'nosniff');
 		}
 
-		// Strict-Transport-Security
-		if ($this->app->get('force_ssl', 0) === 2)
-		{
-			$this->app->setHeader('Strict-Transport-Security', 'max-age=31536000');
```
Comment (wilsonge) [inline plugins/system/httpheader/httpheader.php]: Are we deliberately removing this again?

Comment (zero-24) [reply plugins/system/httpheader/httpheader.php]: based on Brians feedback that we can't use that setting. You can still set it yourself using the subform. Also this should be a informed decision to set that header IMO.

Comment (brianteeman) [reply plugins/system/httpheader/httpheader.php]: My feedback was only that we cant use the value of force_ssl

Comment (zero-24) [reply plugins/system/httpheader/httpheader.php]: yes that is what i mean ;)

**Comment by wilsonge** [comment]
> no we can't use the force HTTPS value in global config. On a good server you do not (and should not) use that

@brianteeman Since when? I was under the impression this should be enabled if you're in SSL mode?

**Comment by mbabker** [comment]
If you're using the web server to force SSL you don't need to turn on the option in the CMS config.

**Comment by wilsonge** [comment]
But it doesn't hurt right? I mean I don't see the point of adding another parameter for this when we have one. Even if you're doing it at .htaccess level it just doesn't change anything

**Comment by zero-24** [comment]
hmm isn't the bigger point the "informed decision" as this is not just a enable that header or not decision. As this can also result into implications that can come after enable that header.

**Comment by brianteeman** [comment]
If you are doing it at htaccess level then you cannot use the if statement - that is all i was saying. I was not saying that you cannot use Strict-Transport-Security or that you shouldnt use it. Just that you cannot rely on checking for forcessl

**Comment by zero-24** [comment]
Ok just combine both notes we keep the `Strict-Transport-Security` out of the default handling. So everyone can do a informed decision on this one.

Can we move forward now? What other points we need to checkout / fix / correct? Or can we start testing?

**Comment by AndyGaskell** [comment]
Nice plug-in, that's cool, thanks @zero-24 

I was wondering, if people think it'd be useful to provide granular CSP support.  This is best explained by the Mozilla docs at https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP

I was at a great talk at the ScotlandPHP conference by Matt Brunt (@Brunty) about how to implement CSP, so I was tinkering with these options on the way home.

**Comment by zero-24** [comment]
> I was wondering, if people think it'd be useful to provide granular CSP support. This is best explained by the Mozilla docs at https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP

Did you try the plugin? What kind of granular CSP support do you want to see? Thanks.

**Comment by rdeutz** [comment]
@zero-24 could you have a look at the code style issues, thanks

**Comment by zero-24** [comment]
Drone is happy now @rdeutz

**Comment by AndyGaskell** [comment]
Hi @zero-24 
Sorry for the slow reply, re...
`Did you try the plugin? What kind of granular CSP support do you want to see? Thanks.`
...Yea, I tried the plug-in, it's cool, nice and simple to set-up.

I was thinking about all content types having a separate setting.

This is pretty much based on a site admin work-flow like...
1. Set the site to `content="default-src 'self'; `
2. See what's broken
3. Fix broken things one at a time, like  `content="default-src 'self'; script-src 'self' *.google-analytics.com "`  to let google analytics work.

In practice an admin would probably do this with `Content-Security-Policy-Report-Only` and just review the reports.

The simplest way to explain is maybe via a screen grab of a a plug-in I wrote...
options screen
...this is from...
https://github.com/AndyGaskell/joomla_plugin_contentsecuritypolicy
...I realise this is quite a different approach, more control and flexibility, but maybe too complex.  

I'm not sure what the best solution is, but thought it might be helpful to add it to the discussion re CSP implementation in Joomla.

Are you heading to JWC?  If you are, be happy to chat about it there.

**Comment by zero-24** [comment]
> Are you heading to JWC? If you are, be happy to chat about it there.

No. please contact me using Glip 😄

**Comment by Quy** [comment]
Installed with your branch and got this error:
```
Fatal error: Declaration of PlgSystemHttpHeader::getSubscribedEvents() must be compatible with Joomla\Event\SubscriberInterface::getSubscribedEvents(): array in C:\xampp\htdocs\joomla-cms-4.0-dev\plugins\system\httpheader\httpheader.php on line 20
```

**Comment by zero-24** [comment]
@Quy should be fixed with the last commit.

**Comment by wilsonge** [comment]
Let's see how this goes for now :)

**Comment by zero-24** [comment]
❤️

**Comment by brianteeman** [comment]
Is there a reason why all the language strings are in the .sys.ini language file and not spread across two language files as normal

**Comment by zero-24** [comment]
see: #19449 @brianteeman

**Comment by Timeforsmilin** [comment]
Hello, I am not a coder, but you guys have hidden the csp and x-frame policies, which are creating duplicate entries, when someone uses alternate security software to protect a Joomla domain. Your csp entries are considered insecure, and may be creating a conflict that disables the administrator save/save&close/close button functions.

I haven't had my shared hosting website taken offline since installing this software https://securitycheck.protegetuordenador.com about 4 years ago. That is more than I can say for any Joomla development! I don't know if my sites were hacked from the shared hosting platform or not, but they have been fine since I have added that security.

Whatever you have created, please make it visible and adjustable, or remove it, as there are better options.

Regards,
Louis

**Comment by ghost** [comment]
@Timeforsmilin can you please open a new Issue? Comments on closed Issues didn't get much Notice.

**Comment by dialanerd** [comment]
What would cause this plugin to stop working?
Everything is enabled and it's the latest version.
None of the code is showing up on my website
I've uninstalled and reinstalled the plugin, but still no luck

Using Joomla 3 (to be upgraded), PHP8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18301">issues.joomla.org/tracker/joomla-cms/18301</a>.</sub>

**Comment by zero-24** [comment]
This here is the J4 version of the plugin this can not be installed into 3.x please use https://github.com/zero-24/plg_system_httpheader within J3.

After activation the plugin configuration has to be adjusted to meet your requirements feel free to follow: https://github.com/zero-24/plg_system_httpheader/blob/master/README.md

**Comment by dialanerd** [comment]
I'm already using use https://github.com/zero-24/plg_system_httpheader but it's not working and I don't understand why?

**Comment by dialanerd** [comment]
I'm not seeing any of the plug-in's settings in my headers.

**Comment by zero-24** [comment]
Ok please create an issue on that repo and include some more details about what you have done, how the plugin configuration looks right now, what is the URL and what you expect. This is the repo for the J4 implementation. Thanks.

**Comment by zero-24** [comment]
https://github.com/zero-24/plg_system_httpheader/issues please

**Pull request merged**